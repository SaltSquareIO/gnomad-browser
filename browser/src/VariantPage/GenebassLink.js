import PropTypes from 'prop-types'
import React from 'react'
// import styled from 'styled-components'

import { ExternalLink, ListItem } from '@gnomad/ui'

import Delayed from '../Delayed'
import { BaseQuery } from '../Query'

const getTopPheno = genebassVariant => {
  return genebassVariant.phenotypes.sort((a, b) => parseFloat(a.pval) - parseFloat(b.pval))[0]
    .analysis_id
}

const genebassQuery = `
query GenebassVariant($variantId: String!) {
  genebass_variant(variantId: $variantId) {
    gene_id
    phenotypes {
      pval
      beta
      analysis_id
    }
  }
}
`

const GenebassLink = ({ variantId }) => {
  return (
    <BaseQuery
      query={genebassQuery}
      variables={{
        variantId,
      }}
    >
      {({ data, error, graphQLErrors, loading }) => {
        let itemContent = null
        if (loading) {
          itemContent = <Delayed>Searching Genebass...</Delayed>
        } else if (error) {
          itemContent = 'Error searching Genebass'
        } else if (!data) {
          if (
            graphQLErrors &&
            graphQLErrors.some(err => err.message === 'Variant not found in Genebass')
          ) {
            itemContent = 'Variant not found in Genebass'
          } else {
            itemContent = (
              <>
                {graphQLErrors && graphQLErrors.length
                  ? Array.from(new Set(graphQLErrors.map(e => e.message))).join(', ')
                  : 'Unable to load variant'}
              </>
            )
          }
        } else {
          const genebassVariant = data.genebass_variant

          if (genebassVariant && genebassVariant.phenotypes) {
            itemContent = (
              <ListItem>
                <ExternalLink
                  href={`https://genebass.org/gene/${
                    genebassVariant.gene_id
                  }/phenotype/${getTopPheno(
                    genebassVariant
                  )}/variant/${variantId}?burdenSet=pLoF&phewasOpts=1&resultIndex=variant-phewas&resultLayout=half`}
                >
                  Genebass variant associations ({genebassVariant.phenotypes.length} P-values)
                </ExternalLink>
              </ListItem>
            )
          } else if (genebassVariant && genebassVariant.gene_id) {
            itemContent = (
              <ListItem>
                <ExternalLink
                  href={`https://genebass.org/gene/${genebassVariant.gene_id}/phenotype/continuous-30690-both_sexes--irnt/variant/${variantId}?resultIndex=variant-phewas&resultLayout=hidden`}
                >
                  Genebass variant found but no associations
                </ExternalLink>
              </ListItem>
            )
          } else {
            itemContent = <ListItem>Variant not found in Genebass</ListItem>
          }
        }

        return <React.Fragment>{itemContent}</React.Fragment>
      }}
    </BaseQuery>
  )
}

GenebassLink.propTypes = {
  variantId: PropTypes.string.isRequired,
}

export default GenebassLink
