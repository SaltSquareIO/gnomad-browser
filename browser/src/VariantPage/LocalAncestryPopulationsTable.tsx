import React from 'react'

import { Badge } from '@gnomad/ui'

import {
  GNOMAD_POPULATION_NAMES,
  LOCAL_ANCESTRY_NAMES,
  PopulationId,
} from '@gnomad/dataset-metadata/gnomadPopulations'

import { PopulationsTable } from './PopulationsTable'

import { LocalAncestryPopulation } from './VariantPage'

type PopulationWithLocalAncestryPopulations = {
  id: PopulationId
  ac: number
  an: number
  subpopulations: LocalAncestryPopulation[]
}

const addPopulationNames = (populations: PopulationWithLocalAncestryPopulations[]) => {
  return populations.map((pop) => ({
    ...pop,
    name: GNOMAD_POPULATION_NAMES[pop.id] || pop.id,

    subpopulations: pop.subpopulations.map((subPop) => ({
      ...subPop,
      name: (LOCAL_ANCESTRY_NAMES as Record<string, string>)[subPop.id.split('_')[1]] || subPop.id,
    })),
  }))
}

const groupPopulations = (
  populations: LocalAncestryPopulation[]
): PopulationWithLocalAncestryPopulations[] => {
  const populationsById: Partial<Record<PopulationId, PopulationWithLocalAncestryPopulations>> = {}

  populations.forEach((pop) => {
    const popId = pop.id.split('_')[0] as PopulationId

    if (!populationsById[popId]) {
      populationsById[popId] = {
        id: popId,
        ac: 0,
        an: 0,
        subpopulations: [],
      }
    }

    populationsById[popId]!.ac += pop.ac
    populationsById[popId]!.an += pop.an
    populationsById[popId]!.subpopulations.push(pop)
  })

  return Object.values(populationsById)
}

type LocalAncestryPopulationsTableProps = {
  populations: LocalAncestryPopulation[]
}

const LocalAncestryPopulationsTable = ({ populations }: LocalAncestryPopulationsTableProps) => {
  const renderedPopulations = addPopulationNames(groupPopulations(populations))

  return (
    <div>
      <PopulationsTable
        populations={renderedPopulations}
        initiallyExpandRows
        showHemizygotes={false}
        showHomozygotes={false}
      />
      <p>
        <Badge level="info">Note</Badge> Local ancestry is not available for all gnomAD genetic
        ancestry groups.
      </p>
    </div>
  )
}

export default LocalAncestryPopulationsTable
