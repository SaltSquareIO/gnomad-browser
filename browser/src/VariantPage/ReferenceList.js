import PropTypes from 'prop-types'
import React from 'react'

import { ExternalLink, List, ListItem } from '@gnomad/ui'

import GenebassLink from './GenebassLink'

export const ReferenceList = ({ variant }) => {
  const ucscReferenceGenomeId = variant.reference_genome === 'GRCh37' ? 'hg19' : 'hg38'
  const { chrom, pos, ref } = variant
  const ucscURL = `https://genome.ucsc.edu/cgi-bin/hgTracks?db=${ucscReferenceGenomeId}&highlight=${ucscReferenceGenomeId}.chr${chrom}%3A${pos}-${
    pos + (ref.length - 1)
  }&position=chr${chrom}%3A${pos - 25}-${pos + (ref.length - 1) + 25}`

  return (
    <List>
      {(variant.rsids || []).length === 1 && (
        <ListItem>
          <ExternalLink
            href={`https://www.ncbi.nlm.nih.gov/projects/SNP/snp_ref.cgi?rs=${variant.rsids[0]}`}
          >
            dbSNP ({variant.rsids[0]})
          </ExternalLink>
        </ListItem>
      )}
      {(variant.rsids || []).length > 1 && (
        <ListItem>
          dbSNP (
          {variant.rsids
            .map(rsid => (
              <ExternalLink
                key={rsid}
                href={`https://www.ncbi.nlm.nih.gov/projects/SNP/snp_ref.cgi?rs=${rsid}`}
              >
                {rsid}
              </ExternalLink>
            ))
            .reduce((acc, el) => [...acc, ', ', el], [])
            .slice(1)}
          )
        </ListItem>
      )}
      <ListItem>
        <ExternalLink href={ucscURL}>UCSC</ExternalLink>
      </ListItem>
      {variant.clinvar && (
        <ListItem>
          <ExternalLink
            href={`https://www.ncbi.nlm.nih.gov/clinvar/variation/${variant.clinvar.clinvar_variation_id}/`}
          >
            ClinVar ({variant.clinvar.clinvar_variation_id})
          </ExternalLink>
        </ListItem>
      )}
      {variant.caid && (
        <ListItem>
          <ExternalLink
            href={`https://reg.clinicalgenome.org/redmine/projects/registry/genboree_registry/by_canonicalid?canonicalid=${variant.caid}`}
          >
            ClinGen Allele Registry ({variant.caid})
          </ExternalLink>
        </ListItem>
      )}
      {variant.reference_genome === 'GRCh38' && (
        <ListItem>
          <ExternalLink
            href={`https://primad.basespace.illumina.com/variant/${variant.variant_id}`}
          >
            primAD
          </ExternalLink>
        </ListItem>
      )}
      {variant.reference_genome === 'GRCh38' && <GenebassLink variantId={variant.variant_id} />}
    </List>
  )
}

ReferenceList.propTypes = {
  variant: PropTypes.shape({
    variant_id: PropTypes.string.isRequired,
    reference_genome: PropTypes.oneOf(['GRCh37', 'GRCh38']).isRequired,
    chrom: PropTypes.string.isRequired,
    pos: PropTypes.number.isRequired,
    ref: PropTypes.string.isRequired,
    caid: PropTypes.string,
    rsids: PropTypes.arrayOf(PropTypes.string),
    clinvar: PropTypes.shape({
      clinvar_variation_id: PropTypes.string.isRequired,
    }),
  }).isRequired,
}
