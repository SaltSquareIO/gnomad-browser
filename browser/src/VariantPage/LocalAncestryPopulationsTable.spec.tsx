import React from 'react'
import { expect, test } from '@jest/globals'
import renderer from 'react-test-renderer'
import { LocalAncestryPopulation } from './VariantPage'

import LocalAncestryPopulationsTable from './LocalAncestryPopulationsTable'

describe('local ancestry populations table', () => {
  test(`has no unexpected changes`, () => {
    const localAncestryData: LocalAncestryPopulation[] = [
      { id: 'amr_african', ac: 1, an: 2 },
      { id: 'amr_amerindigenous', ac: 3, an: 4 },
      { id: 'amr_european', ac: 4, an: 5 },
    ]
    const tree = renderer.create(<LocalAncestryPopulationsTable populations={localAncestryData} />)
    expect(tree).toMatchSnapshot()
  })
})
