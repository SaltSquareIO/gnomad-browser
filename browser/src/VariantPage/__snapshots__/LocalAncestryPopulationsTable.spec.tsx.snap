// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`local ancestry populations table has no unexpected changes 1`] = `
<div>
  <table
    className="Table__BaseTable-sc-7fgtt2-0 PopulationsTable__Table-sc-sijbl0-0 ggAkBv"
  >
    <thead>
      <tr>
        <th
          aria-sort="none"
          colSpan={2}
          scope="col"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            Genetic Ancestry Group
          </button>
        </th>
        <th
          aria-sort="none"
          className="right-align"
          scope="col"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            <span
              className="TooltipHint-sc-1wwbc8j-0 cXXQWi"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              Allele Count
            </span>
          </button>
        </th>
        <th
          aria-sort="none"
          className="right-align"
          scope="col"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            <span
              className="TooltipHint-sc-1wwbc8j-0 cXXQWi"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              Allele Number
            </span>
          </button>
        </th>
        <th
          aria-sort="descending"
          scope="col"
          style={
            {
              "paddingLeft": "25px",
            }
          }
        >
          <button
            onClick={[Function]}
            type="button"
          >
            <span
              className="TooltipHint-sc-1wwbc8j-0 cXXQWi"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              Allele Frequency
            </span>
          </button>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          colSpan={1}
          rowSpan={4}
          scope="row"
        >
          <button
            className="Button__TextButton-sc-1eobygi-3 PopulationsTable__TogglePopulationButton-sc-sijbl0-1 ckJgLT"
            onClick={[Function]}
            type="button"
          >
            Admixed American
          </button>
        </th>
        <td>
          Overall
        </td>
        <td
          className="right-align"
        >
          8
        </td>
        <td
          className="right-align"
        >
          11
        </td>
        <td
          style={
            {
              "paddingLeft": "25px",
            }
          }
        >
          0.7273
        </td>
      </tr>
      <tr
        className="border"
      >
        <td>
          European
        </td>
        <td
          className="right-align"
        >
          4
        </td>
        <td
          className="right-align"
        >
          5
        </td>
        <td
          style={
            {
              "paddingLeft": "25px",
            }
          }
        >
          0.8000
        </td>
      </tr>
      <tr>
        <td>
          Amerindigenous
        </td>
        <td
          className="right-align"
        >
          3
        </td>
        <td
          className="right-align"
        >
          4
        </td>
        <td
          style={
            {
              "paddingLeft": "25px",
            }
          }
        >
          0.7500
        </td>
      </tr>
      <tr>
        <td>
          African
        </td>
        <td
          className="right-align"
        >
          1
        </td>
        <td
          className="right-align"
        >
          2
        </td>
        <td
          style={
            {
              "paddingLeft": "25px",
            }
          }
        >
          0.5000
        </td>
      </tr>
    </tbody>
    <tfoot>
      <tr
        className="border"
      >
        <th
          colSpan={2}
          scope="row"
        >
          Total
        </th>
        <td
          className="right-align"
        >
          8
        </td>
        <td
          className="right-align"
        >
          11
        </td>
        <td
          style={
            {
              "paddingLeft": "25px",
            }
          }
        >
          0.7273
        </td>
      </tr>
    </tfoot>
  </table>
  <p>
    <span
      className="Badge__BadgeWrapper-sc-j4izdp-1 jEnuJq"
    >
      Note
    </span>
     Local ancestry is not available for all gnomAD genetic ancestry groups.
  </p>
</div>
`;
